### ISSUE TITLE
HTTP Toolkit Pro - Unable to Intercept Chrome Traffic on Android

### ISSUE THREAD

**Issue by H0r53** [httptoolkit/httptoolkit#260]
I purchased HTTP Toolkit pro and installed it on both my Macbook and a rooted android phone. I am able to successfully capture the traffic from apps on the phone. However, I am not able to intercept any traffic from Chrome. In fact, I can't even access any sites in Chrome when HTTP Toolkit is running. I've installed the HTTP Toolkit certificate on the Android phone as a user certificate and used a Magisk module called "TrustAllUserCerts" which basically moves the user certs to system certs on boot.

How can I intercept Chrome / Chromium traffic with HTTP Toolkit? It's kinda crazy that it works on apps but not on Chrome itself. The documentation [here](https://httptoolkit.tech/docs/guides/android/) says it should automatically work with Chrome.

**Comment by pimterry** [comment]
Hmm, that's very unusual, I've never seen any device where Chrome's certificate trust was stricter than other applications - I don't know of any reason why that would happen.

Just to confirm, when you connect the HTTP Toolkit Android app, it shows both user & system certificates as trusted, is that right?

If the device is rooted, you shouldn't need to install the user certificate or use that Magisk module at all. What happens if you disable that Magisk module? Is there anything else unusual about your device or the version of Chrome you're using?

What is shown on the View page of HTTP Toolkit on your computer when you're trying to browse with Chrome?

**Comment by H0r53** [comment]
> Just to confirm, when you connect the HTTP Toolkit Android app, it shows both user & system certificates as trusted, is that right?
Yes, a green checkmark is next to both.

> What happens if you disable that Magisk module?
After disabling the Magisk module and reboot. HTTP Toolkit on the phone now only shows a green check mark next to "User Trust Enabled" and there is a warning sign suggesting "System Trust Enabled" is not working. More specifically, HTTPS interception will only be available for apps that are configured to trust user installed certificates. Whenever I "Test Interception" I'm taken to a Chrome tab that says "No internet"

>  Is there anything else unusual about your device or the version of Chrome you're using?
No, it's just a freshly rooted Pixel 2 using the default version of Google Chrome

> What is shown on the View page of HTTP Toolkit on your computer when you're trying to browse with Chrome?
The view page show some successful intercepted requests to `android.httptoolkit.tech`, `update.googleapis.com`, `ssl.google-analytics.com`, and `play.googleapis.com`. But there are only around 6 total and none of this traffic is coming from Chrome.

**Comment by pimterry** [comment]
> The view page show some successful intercepted requests to android.httptoolkit.tech, update.googleapis.com, ssl.google-analytics.com, and play.googleapis.com. But there are only around 6 total and none of this traffic is coming from Chrome.

Are there no lines with messages like "Certificate rejected" or "Connected reset"? Messages like this:

![Screenshot from 2022-03-28 16-56-39](https://user-images.githubusercontent.com/1526883/160426641-04b9255f-0984-4e5c-8a2e-046a617f0f2d.png)

If the device is connecting to your computer but it doesn't trust the certificate, you should see warnings like this about the certificate being refused on your computer.

If you don't see any warnings, then it's a different problem: the traffic isn't being sent to HTTP Toolkit at all. That would be even stranger, and suggests maybe there's some independent internet access detection that's being tripped and blocking these connections completely.

I think there may be some general issues with more recent Magisk versions which might be related to this (https://github.com/httptoolkit/httptoolkit-android/issues/8). I'm setting up a device to reproduce those at the moment - can you please share the exact steps & versions you used to root your device?

**Comment by H0r53** [comment]
I'm using 
- OS: Android 11; Pixel 2 XL Build/RP1A.201005.004.A1
- Image: taimen-rp1a.201005.004.a1
- Magisk-v23.0.apk
- MagiskManager-v5.4.1.apk
- Chrome version on Pixel 99.0.4744.73 
- HTTP Toolkit version on phone: 1.3.2
- HTTP Toolkit version on Mac Host: 1.7.0

**Comment by H0r53** [comment]
> Are there no lines with messages like "Certificate rejected" or "Connected reset"? 

No, and the lines that do appear are not coming from Chrome. They are coming from some other service being intercepted, my guess would be default Android services.

 The Chrome error specifically is `No Internet` - `ERR_PROXY_CONNECTION_FAILED`. It sounds like the Chrome proxy is not configured correctly? Can I change settings in Chrome to establish a successful connection to the HTTP Toolkit proxy?

**Comment by H0r53** [comment]
Also, if it helps, HTTP Toolkit on the host shows "Connected Sources" as the following:

- GoogleAnalytics/220.6.18 (Android 11)
- Okhttp/4.3.0

Should I change the android wifi adapter to use the HTTP Toolkit proxy settings?

**Comment by H0r53** [comment]
@pimterry the situation seems to have evolved. I completely removed the HTTP Toolkit certificate from the phone then reinstalled it, rebooted, etc. I can now see Chrome traffic from the phone within the HTTP Toolkit `View` section on the host. However, I'm consistently receiving "Certificate rejected for connection to ______" where ____ is whatever resource Chrome is attempting to access. I'm also seeing some "Aborted connection to _____" errors as well. Intercepting traffic from apps is still successful.

**Comment by H0r53** [comment]
I tested on a different machine and I've made some progress on this (I can provide more details later) and now I'm able to successfully capture all traffic **except for android app traffic using a WebView**. I'm under the impression WebViews basically use Chromium to render a webpage, and if I can capture traffic in Chrome I should be able to capture WebView traffic. Is it possible SSL Pinning, or something similar, is occurring within an Android application's WebView? I'm already injecting the HTTP Toolkit certificate with Frida, which is required to intercept the standard traffic (I believe from http3). The behavior I'm seeing seems to indicate either Chromium isn't respecting the HTTP Toolkit certificate or the WebView is performing another form of SSL Pinning. Since I can intercept traffic from Chrome, I'm leaning towards the latter. Any insights on how SSL Pinning may be used for WebViews and how to bypass?

**Comment by H0r53** [comment]
Issue resolved - kinda. My Macbook host uses an upstream proxy / certificate authority. I had already imported the root CA cert into HTTP Toolkit, however, I needed to import that cert into the Android certificate store as well. The reason is, even if the HTTP Toolkit certificate is installed, by proxying into HTTP Toolkit on my mac host, which then proxies through another service, the upstream proxy certificate is not trusted by the phone itself. This causes issues with Chrome and other apps. SSL Pinning was an obstacle in my case, but it didn't really have anything to do with HTTP Toolkit.

The reason this is not fully resolved is I cannot reliably get HTTP Toolkit to work on my Macbook (using the upstream proxy). I managed to somehow get it to work earlier today, but I ran into the SSL pinning issue and decided to try on another machine. After eventually finding success on the other machine, I switched back to my Macbook and once again I get certificate issues when attempting to test the HTTP Toolkit connection in Chrome.

What is the appropriate way to use HTTP Toolkit when the network has a proxy at the edge?

**Comment by pimterry** [comment]
Wow, ok, there's a lot going on here!

By any chance is this device managed by a enterprise policy or Android work account? Or is it running on a network that might be using WPAD to inject proxy config? If there is an enterprise/network configuration somewhere that's setting its own proxy settings that Chrome is using automatically, and those settings are conflicting with HTTP Toolkit's settings, then I think that could explain some of this behaviour.

You also mentioned HTTP/3 - is there a specific reason for that? HTTP Toolkit doesn't yet support HTTP/3, but that hasn't been an issue so far as not supporting HTTP/3 is very common, and every implementation I've seen happily falls back to HTTP/2 or HTTP/1 when an HTTP proxy is configured explicitly.

> I'm under the impression WebViews basically use Chromium to render a webpage, and if I can capture traffic in Chrome I should be able to capture WebView traffic. Is it possible SSL Pinning, or something similar, is occurring within an Android application's WebView?

I believe WebViews will use the app's [network security configuration](https://developer.android.com/training/articles/security-config.html), so I think the certs they trust will match the configuration of the app itself, not Chrome. By default that configuration only trusts system certificates, not user certificates, and there are options that can be set there to pin specific certificates too.

I'm not actually sure exactly how that interacts with Frida. It's possible that WebViews implementation is somewhat isolated from the app, and so the Frida script doesn't work, in which case you'd be able to see the app traffic but not the traffic from the WebView within - is that what's happening? If so it might be interesting to try [apk-mitm](https://github.com/shroudedcode/apk-mitm) which can take an APK and rewrite the network security configuration entirely. In theory that should unpin even webviews from every app you like.

I haven't heard about this before, but it sounds plausible, do let me know if that's what's happening here.

> The reason is, even if the HTTP Toolkit certificate is installed, by proxying into HTTP Toolkit on my mac host, which then proxies through another service, the upstream proxy certificate is not trusted by the phone itself.

As far as I'm aware, that shouldn't be possible. For all connections that HTTP Toolkit redirects, the TLS connection is fully terminated by HTTP Toolkit. It should be impossible proxy traffic through HTTP Toolkit and ever see an upstream certificate (if you did, that direct connection would be an encrypted tunnel that HTTP Toolkit wouldn't be able to inspect or modify, and avoiding that is the whole point).

Of course, given that it's working something else must be happening here, but it shouldn't be. 

It might be interesting to run Wireshark on your computer. That won't show you encrypted data, but you should be able to see the raw successful and failing TLS connections there, and there might be some clues in that.

> What is the appropriate way to use HTTP Toolkit when the network has a proxy at the edge?

In general, to use an upstream proxy all that's required is to set the upstream proxy details in the "Connection Settings" section of HTTP Toolkit's settings, and add the proxy's CA certificate there too (under 'Trusted CA Certificates'). That should be all that's necessary, and nothing should be required on downstream devices at all.

**Comment by H0r53** [comment]
> By any chance is this device managed by a enterprise policy or Android work account? Or is it running on a network that might be using WPAD to inject proxy config? 

My Macbook uses a service that basically acts as a VPN/Proxy. I have a CA file for the proxy and I generally have to import it to get certain browsers and package managers to work.

> You also mentioned HTTP/3

I don't recall mentioning HTTP/3. It is out of scope of my current research and it hasn't been a problem related to anything in this ticket.

>  It's possible that WebViews implementation is somewhat isolated from the app, and so the Frida script doesn't work, in which case you'd be able to see the app traffic but not the traffic from the WebView within - is that what's happening?

The WebView in this case was using a form of SSL pinning, which I've seen before but never in a WebView. I managed to bypass it after research. This was not a HTTP Toolkit issue. I don't imagine HTTP Toolkit can do anything about app-level SSL pinning.

> As far as I'm aware, that shouldn't be possible. For all connections that HTTP Toolkit redirects, the TLS connection is fully terminated by HTTP Toolkit. 

I could have misdiagnosed this. I did experience the same `ERR_PROXY_CONNECTION_FAILED` issue on the Android when using HTTP Toolkit on a different machine (no upstream proxy). However, it was much easier to eventually resolve on the different machine (Scan QR code, then try the ADB connection, eventually it worked). My thoughts were the phone is proxied into HTTP Toolkit on my mac, then traffic leaving HTTP Toolkit is proxied through another service at the network edge. So it seemed to make sense that the phone was receiving connections signed by the upstream proxy and thus causing the errors I've experienced. However, if the SSL connection is fully terminated between the phone and HTTP Toolkit on my Macbook, then restablished between HTTP Toolkit and the upstream proxy, then I agree it doesn't make sense for this to be the issue.

All I know is, after installing the upstream proxy certificate on the phone, alongside the HTTP Toolkit cert, it eventually worked and I could capture all traffic (both from apps and Chrome).

> In general, to use an upstream proxy all that's required is to set the upstream proxy details in the "Connection Settings" section of HTTP Toolkit's settings, and add the proxy's CA certificate there too (under 'Trusted CA Certificates'). That should be all that's necessary, and nothing should be required on downstream devices at all.

That makes sense, and that's what I did from the beginning. But I still had issues and I'm still having them. I can use the alternative computer setup for my current needs, but something very strange is still occurring.

**Comment by H0r53** [comment]
@shirshak55 what do you mean by change the gateway to an invalid IP? Are you suggesting to setup the Android WiFi connection such that it uses a proxy, but provide an invalid proxy IP?

**Comment by pimterry** [comment]
I'm pretty sure I've found and fixed the root cause of this. I wrote up a blog post on the details here: https://httptoolkit.tech/blog/chrome-android-certificate-transparency/.

This was an update in Chrome 99 that in rooted use cases broke traffic interception for Chrome specifically, while it kept working for other apps, as you're describing. It's not totally clear if that's the whole story from the details here, but you do mention `ERR_CERTIFICATE_TRANSPARENCY_REQUIRED` in places, which definitely relates to this.

This is now fixed in the latest HTTP Toolkit, which will automatically patch some settings in Chrome to avoid this issue. Your HTTP Toolkit will update in the background next time you open it, and then this should work correctly the following time.

Thanks for reporting this! I'm going to close this issue now since I think this is very likely to have been the cause, but do feel free to reply if you're still seeing issues after the latest HTTP Toolkit update.
